### PR TITLE
mgr/dashboard: fix set-ssl-certificate{,-key} commands

### PR DESCRIPTION
**Pull request by alfonsomthd**
- Now create-self-signed-cert command relies on set-ssl-certificate{,-key} commands.
- Simplify the command testing and increase the test coverage.

Fixes: https://tracker.ceph.com/issues/50519
Signed-off-by: Alfonso Martínez <almartin@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by nizamial09 — APPROVED** [review]

**Review by avanthakkar — APPROVED** [review]

File: src/pybind/mgr/dashboard/module.py [hunk 1]
```diff
@@ -360,28 +366,26 @@ def shutdown(self):
...
-                'Please specify the certificate key file with "-i" option'
+            return -errno.EINVAL, '', self.MSG_SSL_CERT_KEY_ERROR
         if mgr_id is not None:
-            self.set_store(_get_localized_key(mgr_id, 'key'), inbuf.decode())
+            self.set_store(_get_localized_key(mgr_id, 'key'), inbuf)
         else:
-            self.set_store('key', inbuf.decode())
-        return 0, 'SSL certificate key updated', ''
+            self.set_store('key', inbuf)
+        return 0, self.MSG_SSL_CERT_KEY_UPDATED, ''
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/module.py:388]: I see these 2 (or 3 actually) functions are exactly the same except for some part in the error message (`certificate file` vs `certificate key`) and the key name... So, what about a parameterized function:
```py
# Literal may need typing_extensions in 3.6
def set_ssl_item(self, item_label: str, item_key: Literal['crt', 'key'], mgr_id: Optional[str]):
  if inbuf is None:
    return -errno.EINVAL, '', 'Please specify the {} with "-i" option'.format(item_label)

  if mgr_id is not None:
    self.set_store(_get_localized_key(mgr_id, item_key), inbuf)
  else:
    self.set_store(item_key, inbuf)
  return 0, 'SSL {item} updated'.format(item_label), ''

@CLIWriteCommand("dashboard set-ssl-certificate")
def set_ssl_certificate(self, mgr_id: Optional[str] = None, inbuf: Optional[str] = None):
  return self.set_ssl_item('certificate', 'crt', mgr_id)

@CLIWriteCommand("dashboard set-ssl-certificate-key")
def set_ssl_certificate(self, mgr_id: Optional[str] = None, inbuf: Optional[str] = None):
  return self.set_ssl_item('certificate key', 'key', mgr_id)

@CLIWriteCommand("dashboard create-self-signed-cert")
def create_self_signed_cert(self):
  cert, pkey = create_self_signed_cert('IT', 'ceph-dashboard')
  r = HandleCommandResult(*self.set_ssl_item('self-signed certificate', 'crt', mgr_id=None))
  if r.retval != 0:
    return r

  r = HandleCommandResult(*self.set_ssl_item('self-signed certificate key', 'key', mgr_id=None))
  if r.retval != 0:
    return r
  return 0, Self-signed certificate created', ''
```

Comment (epuertat) [reply src/pybind/mgr/dashboard/module.py:388]: The more I see these functions (which require the Dashboard to be running) the more I'm convinced that we should remove all the dummy commands that can be easily replaced with a `ceph config set...` (which is something that can be run before starting the dashboard) :/

File: src/pybind/mgr/dashboard/tests/__init__.py [hunk 2]
```diff
@@ -38,16 +49,17 @@ def exec_dashboard_cmd(command_handler, cmd, **kwargs):
...
-            return out
-
-    ret, out, err = CLICommand.COMMANDS[cmd_dict['prefix']].call(mgr, cmd_dict, inbuf)
+    else:
+        dashboard_module = ModuleTestClass()
+        if cmd_dict['prefix'] in CLICommand.COMMANDS:
+            ret, out, err = CLICommand.COMMANDS[cmd_dict['prefix']].call(dashboard_module, cmd_dict,
+                                                                         inbuf)
+        else:
+            ret, out, err = dashboard_module.handle_command(inbuf, cmd_dict)
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/tests/__init__.py:61]: Why is this needed?

File: src/pybind/mgr/dashboard/module.py [hunk 3]
```diff
@@ -409,8 +413,8 @@ def handle_command(self, inbuf, cmd):
 
     def create_self_signed_cert(self):
         cert, pkey = create_self_signed_cert('IT', 'ceph-dashboard')
-        self.set_store('crt', cert)
-        self.set_store('key', pkey)
+        self.set_ssl_certificate(inbuf=cert)
+        self.set_ssl_certificate_key(inbuf=pkey)
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/module.py:417]: What about removing this command from the legacy `handle_command` switch-case and putting it here? And I think we should also check if the previous sub-command failed.

```py
@CLIWriteCommand("dashboard create-self-signed-cert")
def create_self_signed_cert(self):
  cert, pkey = create_self_signed_cert('IT', 'ceph-dashboard')
  r = HandleCommandResult(*self.set_ssl_item('self-signed certificate', 'crt', mgr_id=None))
  if r.retval != 0:
    return r

  r = HandleCommandResult(*self.set_ssl_item('self-signed certificate key', 'key', mgr_id=None))
  if r.retval != 0:
    return r
  return 0, Self-signed certificate created', ''  
```

File: src/pybind/mgr/dashboard/tests/test_ssl.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            self.exec_cmd('set-ssl-certificate-key', inbuf='', mgr_id='x')
+        self.assertEqual(ctx.exception.retcode, -errno.EINVAL)
+        self.assertEqual(str(ctx.exception), Module.MSG_SSL_CERT_KEY_ERROR)
+
+        result = self.exec_cmd('set-ssl-certificate-key', inbuf='content', mgr_id='x')
+        self.assertEqual(result, Module.MSG_SSL_CERT_KEY_UPDATED)
+
+    def test_create_self_signed_cert(self):
+        result = self.exec_cmd('create-self-signed-cert')
+        self.assertEqual(result, Module.MSG_SSL_SELF_SIGNED_CERT_CREATED)
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/tests/test_ssl.py:29]: new tests!! 🥳

File: src/pybind/mgr/dashboard/module.py (hunk 1 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/module.py:388]: > I see these 2 (or 3 actually) functions are exactly the same except for some part in [...]

Of course more refactoring can be done, but as per our previous agreement in Dashboard Team regarding bugfixes,
my intention here was to provide only the minimal fix covered by tests.

File: src/pybind/mgr/dashboard/tests/__init__.py (hunk 2 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/tests/__init__.py:61]: In order to test dashboard module ```handle_command``` that in turn calls ```create-self-signed-cert``` so we gain test coverage for both (previously ```handle_command``` was not being tested).
Here ```CLICommand.COMMANDS[cmd_dict['prefix']].call(dashboard_module,``` I replaced the mgr mock with dashboard module following the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle).

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-04-30_03:09:53-rados-wip-kefu-testing-2021-04-30-0020-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-30_10:06:45-rados-wip-kefu-testing-2021-04-30-1335-distro-basic-smithi/

File: src/pybind/mgr/dashboard/tests/__init__.py (hunk 2 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/tests/__init__.py]: Simplified the command testing through ```CLICommandTestMixin.exec_cmd```.

File: src/pybind/mgr/dashboard/module.py [hunk 5]
```diff
@@ -14,9 +14,16 @@
...
 import time
-from typing import Optional
+from typing import TYPE_CHECKING, Optional
 
-from mgr_module import CLIWriteCommand, MgrModule, MgrStandbyModule, Option, _get_localized_key
+if TYPE_CHECKING:
+    if sys.version_info >= (3, 8):
+        from typing import Literal
+    else:
+        from typing_extensions import Literal
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/module.py:23]: No need to change, but just as a reference, [this is how other libraries do that](https://github.com/agronholm/typeguard/blob/master/src/typeguard/__init__.py#L26-L32).

**Review by epuertat — APPROVED** [review]

File: src/pybind/mgr/dashboard/module.py (hunk 5 above)
Comment (epuertat) [reply src/pybind/mgr/dashboard/module.py:23]: Thanks @alfonsomthd for addressing my suggestions! :D

Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/module.py:23]: > No need to change, but just as a reference, [this is how other libraries do that](https://github.com/agronholm/typeguard/blob/master/src/typeguard/__init__.py#L26-L32).

I just applied the same approach as [here](https://github.com/ceph/ceph/pull/36146/commits/f002c6ce40338c08dbc91e08682365e07662f135) :grin:
